src/librbd/mirror/snapshot/GroupCreatePrimaryRequest.cc

```diff
@@ -0,0 +1,577 @@
+
+  std::stringstream ind_snap_name_stream;
+  ind_snap_name_stream << ".mirror.primary."
+                       << global_group_id << "." << snap_id;
+  return ind_snap_name_stream.str();
+}
+
+
+template <typename I>
+struct C_ImageSnapshotCreate2 : public Context {
+  I *ictx;
+  uint64_t snap_create_flags;
+  int64_t group_pool_id;
+  std::string group_id;
+  std::string group_snap_id;
+  uint64_t *snap_id;
+  Context *on_finish;
+
+  cls::rbd::MirrorImage mirror_image;
+  mirror::PromotionState promotion_state;
+  std::string primary_mirror_uuid;
+
+  C_ImageSnapshotCreate2(I *ictx, uint64_t snap_create_flags,
+                         int64_t group_pool_id,
+                         const std::string &group_id,
+                         const std::string &group_snap_id,
+                         uint64_t *snap_id,
+                         Context *on_finish)
+    : ictx(ictx), snap_create_flags(snap_create_flags),
+      group_pool_id(group_pool_id), group_id(group_id),
+      group_snap_id(group_snap_id), snap_id(snap_id),
+      on_finish(on_finish) {
+  }
+
+  void finish(int r) override {
+    if (r < 0 && r != -ENOENT) {
+      on_finish->complete(r);
+      return;
+    }
+
+    if (mirror_image.mode != cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT ||
+        mirror_image.state != cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+      lderr(ictx->cct) << "snapshot based mirroring is not enabled" << dendl;
+      on_finish->complete(-EINVAL);
+      return;
+    }
+
+    auto req = mirror::snapshot::CreatePrimaryRequest<I>::create(
+      ictx, mirror_image.global_image_id, CEPH_NOSNAP, snap_create_flags, 0U,
+      group_pool_id, group_id, group_snap_id, snap_id, on_finish);
+    req->send();
+  }
+};
+
+
+template <typename I>
+void image_snapshot_create2(I *ictx, uint32_t flags,
+                            const std::string &group_snap_id,
+                            uint64_t *snap_id, Context *on_finish) {
+  CephContext *cct = ictx->cct;
+  ldout(cct, 10) << "ictx=" << ictx << dendl;
+
+  uint64_t snap_create_flags = 0;
+  int r = librbd::util::snap_create_flags_api_to_internal(cct, flags,
+                                                          &snap_create_flags);
+  if (r < 0) {
+    on_finish->complete(r);
+    return;
+  }
+
+  auto on_refresh = new LambdaContext(
+    [ictx, snap_create_flags, group_snap_id, snap_id, on_finish](int r) {
+      if (r < 0) {
+        lderr(ictx->cct) << "refresh failed: " << cpp_strerror(r) << dendl;
+        on_finish->complete(r);
+        return;
+      }
+
+      auto ctx = new C_ImageSnapshotCreate2<I>(ictx, snap_create_flags,
+                                               ictx->group_spec.pool_id,
+                                               ictx->group_spec.group_id,
+                                               group_snap_id, snap_id,
+                                               on_finish);
+      auto req = mirror::GetInfoRequest<I>::create(*ictx, &ctx->mirror_image,
```

Comment:
This can be done without C_ImageSnapshotCreate2 as this is a separate class .

A function get_info() will call GetInfoRequest() and pass handle_get_info() to the callback.
handle_get_info() will call create_snapshot()  on success. etc etc